File: services/nuldoc/nuldoc-src/pages/TagListPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import BlogGlobalHeader from "../components/BlogGlobalHeader.ts";
import SlidesGlobalHeader from "../components/SlidesGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import { Config } from "../config.ts";
import { TagPage } from "../generators/tag.ts";
import { a, article, elem, Element, footer, h1, h2, header } from "../dom.ts";

export default async function TagListPage(
  tags: TagPage[],
  site: "blog" | "slides",
  config: Config,
): Promise<Element> {
  const pageTitle = "タグ一覧";

  const GlobalHeader = site === "blog" ? BlogGlobalHeader : SlidesGlobalHeader;

  return await PageLayout({
    metaCopyrightYear: config.site.copyrightYear,
    metaDescription: "タグの一覧",
    metaTitle: `${pageTitle}｜${config.sites[site].siteName}`,
    site,
    config,
    children: elem(
      "body",
      { class: "list" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        header({ class: "page-header" }, h1({}, pageTitle)),
        ...Array.from(tags)
          .sort((t1, t2) => {
            const ta = t1.tagSlug;
            const tb = t2.tagSlug;
            if (ta < tb) return -1;
            if (ta > tb) return 1;
            return 0;
          })
          .map((tag) => {
            const posts = tag.numOfPosts === 0
              ? ""
              : `${tag.numOfPosts}件の記事`;
            const slides = tag.numOfSlides === 0
              ? ""
              : `${tag.numOfSlides}件のスライド`;
            const footerText = `${posts}${
              posts && slides ? "、" : ""
            }${slides}`;

            return article(
              { class: "post-entry" },
              a(
                { href: tag.href },
                header(
                  { class: "entry-header" },
                  h2({}, tag.tagLabel),
                ),
                footer({ class: "entry-footer" }, footerText),
              ),
            );
          }),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/tag_list_page.rb
module Nuldoc
  module Pages
    class TagListPage
      def initialize(tags:, site:, config:)
        @tags = tags
        @site = site
        @config = config
      end

      def render
        site = @site
        config = @config
        page_title = 'タグ一覧'
        global_header = site == 'blog' ? Components::BlogGlobalHeader : Components::SlidesGlobalHeader
        site_entry = config.site_entry(site)

        sorted_tags = @tags.sort_by(&:tag_slug)

        Components::PageLayout.new(
          meta_copyright_year: config.site.copyright_year,
          meta_description: 'タグの一覧',
          meta_title: "#{page_title}｜#{site_entry.site_name}",
          site: site,
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'list' do
              render(global_header, config: config)
              main class: 'main' do
                header(class: 'page-header') { h1 { text page_title } }
                sorted_tags.each do |tag|
                  posts_text = tag.num_of_posts.zero? ? '' : "#{tag.num_of_posts}件の記事"
                  slides_text = tag.num_of_slides.zero? ? '' : "#{tag.num_of_slides}件のスライド"
                  separator = !posts_text.empty? && !slides_text.empty? ? '、' : ''
                  footer_text = "#{posts_text}#{separator}#{slides_text}"

                  article class: 'post-entry' do
                    a href: tag.href do
                      header(class: 'entry-header') { h2 { text tag.tag_label } }
                      footer(class: 'entry-footer') { text footer_text }
                    end
                  end
                end
              end
              render(Components::GlobalFooter, config: config)
            end
          end
        ).build
      end
    end
  end
end
